### PR TITLE
Microphysics emulator returns dictionary

### PR DESCRIPTION
**Pull request by nbren12**
The microphysics emulator returns a list rather than a dictionary. Handling the variable name logic makes it hard to extend this class, complicates the internal implementation, and increased the need for custom logic at evaluation time like `_unpack_predictions`. `keras` already supports dictionary output, so this PR refactors the microphysics class to use it. It also adds some supporting public APIs.

Added public API:
- `fv3fit.keras.adapters.{rename_dict_output,convert_to_dict_output}`

Significant internal changes:
- Modify microphysics emulator to return dicts
- add backwards compatible loading tools
- delete tests targeted at implementation details (w/o significant loss of coverage).

### CODE REVIEW

File: external/fv3fit/tests/emulation/models/test_microphysics_config.py [hunk 1]
```diff
@@ -61,105 +61,17 @@ def test_Config_asdict():
     assert result == original
 
 
-def test_Config__processed_inputs():
```
Comment (nbren12) [inline external/fv3fit/tests/emulation/models/test_microphysics_config.py:64]: these tested the helper functions inside of `.build` and were easy to break as I changed their APIs. The coverage is the same after deleting them.

File: external/fv3fit/tests/emulation/test_train_microphysics.py [hunk 2]
```diff
@@ -146,37 +128,6 @@ def test_TrainConfig_from_args_sysargv(monkeypatch):
     assert config.model.architecture.name == "rnn"
 
 
-def test_TrainConfig_invalid_input_vars():
```
Comment (nbren12) [inline external/fv3fit/tests/emulation/test_train_microphysics.py:149]: This is impossible to happen now since the variables are now inferred from the model config.

File: external/fv3fit/tests/keras/test_upstream_apis.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    a = tf.keras.Input(name="a", shape=[5])
+    # make sure that `name`` is ignored
+    b = tf.keras.layers.Dense(3, name="not_b")(a)
+    model = tf.keras.models.Model(inputs=[a], outputs={"b": b})
+
+    a = tf.ones((1, 5))
+    out = model(a)
+    assert set(out) == {"b"}
+
+
```
Comment (frodre) [inline external/fv3fit/tests/keras/test_upstream_apis.py:86]: If I'm understanding correctly, keras models will ignore extra keys in an input dictionary.  Maybe include an upstream test of dictionary inputs with the variation having extra keys.  If I'm mistaken, ignore 😅

**Review by frodre — APPROVED** [review]
Thanks so much for taking this on.  There are a lot of nice simplifications in here that I think will help a lot going forward.  Only minor comments and suggestions. 

I'll set up local data training test to see if there's much of a change on the timing.

File: external/fv3fit/fv3fit/train_microphysics.py [hunk 4]
```diff
@@ -217,22 +182,31 @@ def main(config: TrainConfig, seed: int = 0):
...
+    train_set = next(iter(train_ds.shuffle(100_000).batch(50_000)))
+    test_set = next(iter(test_ds.shuffle(160_000).batch(80_000)))
 
-    model = config.model.build(
-        X_train, sample_direct_out=direct_sample, sample_residual_out=resid_sample
+    model = config.model.build(train_set)
+    config.loss.prepare(
+        output_names=list(
+            set(config.model.direct_out_variables)
+            | set(config.model.residual_out_variables)
```
[the remark this answers is not in the document]
Comment (nbren12) [reply external/fv3fit/fv3fit/train_microphysics.py:192]: Agreed. I have this in my follow-up PR (#1508).

File: external/fv3fit/fv3fit/emulation/keras.py [hunk 5]
```diff
@@ -52,19 +48,9 @@ def score_model(
...
-    elif len(model.output_names) == 1:
-        scores, profiles = score_single_output(
-            targets, prediction, model.output_names[0]
-        )
-    else:
-        logger.error("Tried to call score on a model with no outputs.")
-        raise ValueError("Cannot score model with no outputs.")
-
-    return scores, profiles
+    prediction = model.predict(data)
```
[the remark this answers is not in the document]
Comment (nbren12) [reply external/fv3fit/fv3fit/emulation/keras.py:51]: Don't think so. I could add a test to the  upstream_tests.

File: external/fv3fit/tests/keras/test_upstream_apis.py (hunk 3 above)
Comment (nbren12) [reply external/fv3fit/tests/keras/test_upstream_apis.py:86]: > keras models will ignore extra keys in an input dictionary. 

I'm a bit confused about the request. Is there a problem with ignoring certain variables in the input dictionary?

Comment (frodre) [reply external/fv3fit/tests/keras/test_upstream_apis.py:86]: > Is there a problem with ignoring certain variables in the input dictionary?

Nope no problem at all, just seemed like a good target to test against to make sure it continues to work.

Comment (nbren12) [reply external/fv3fit/tests/keras/test_upstream_apis.py:86]: Sounds good. done.

File: external/fv3fit/fv3fit/train_microphysics.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nbren12) [reply external/fv3fit/fv3fit/train_microphysics.py]: Also, your suggestion above renders this moot.
